### ISSUE TITLE
[Question] Best ACC and Best AUC Never Get Updated in Inference State / Function?

### ISSUE THREAD

**Issue by tim5go** [facebookresearch/dlrm#216]
https://github.com/facebookresearch/dlrm/blob/9c2fda79afbc09e277c17e420bffe510125b4f70/dlrm_s_pytorch.py#L1353-L1357
https://github.com/facebookresearch/dlrm/blob/9c2fda79afbc09e277c17e420bffe510125b4f70/dlrm_s_pytorch.py#L750-L759
https://github.com/facebookresearch/dlrm/blob/9c2fda79afbc09e277c17e420bffe510125b4f70/dlrm_s_pytorch.py#L880-L890

From the above code code snippets, it looks like to me that the values **best_acc_test** and **best_auc_test** never get updated as it only stays in the inference function scope.

Is it designed for that? or if it is actually a bug?

**Comment by mnaumovfb** [comment]
This looks like a bug to me, but luckily the side effect should not be too bad. As far as I can see all the reporting/prints of metrics will be correct, but 
1. The variable `is_best` will always be checked with respect to the initial values of the metrics. Therefore, the model will always be declared to have best accuracy/AUC and will always be checkpointed (when `--save-model` flag is selected). The correct behavior would be to checkpoint only if the model is best across all prior iterations.
2. The early stopping of training when `--mlperf-logging` is used with either `--mlperf-acc-threshold` or `--mlperf-auc-threshold` flags will not be triggered, therefore the training will continue until the end.

We will take that as an action item to fix. Thank you for reporting it.

**Comment by future-xy** [comment]
Yes, I just found this problem too. During the training & testing, the code will always save the current model state, as shown in the following log:
`
Finished training it 14336/306969 of epoch 0, 53.08 ms/it, loss 0.468385
Testing at - 14336/306969 of epoch 0,
 accuracy 77.802 %, best 77.802 %
Saving model to dlrm/output/best_kaggle_model.pt
Finished training it 15360/306969 of epoch 0, 53.15 ms/it, loss 0.466621
Testing at - 15360/306969 of epoch 0,
 accuracy 77.742 %, best 77.742 %
Saving model to dlrm/output/best_kaggle_model.pt
`
